### PR TITLE
chore: Try SSA for Kamelet install only once

### PR DESCRIPTION
**Pull request by astefanutti**
This PR follows up #2814 to take into account https://github.com/apache/camel-k/pull/2814#discussion_r764679710.

@zregvart let me know if that corresponds to what you had in mind.

**Release Note**
```release-note
NONE
```

### CODE REVIEW

File: pkg/install/kamelets.go [hunk 1]
```diff
@@ -50,7 +52,15 @@ const (
...
-var hasServerSideApply = true
+var (
+	log = logf.Log
+
+	hasServerSideApply atomic.Value
+)
+
+func init() {
+	hasServerSideApply.Store(true)
+}
```
Comment (zregvart) [inline pkg/install/kamelets.go:63]:
```suggestion
```

File: pkg/install/kamelets.go [hunk 2]
```diff
@@ -50,7 +52,15 @@ const (
 	defaultKameletDir = "/kamelets/"
 )
 
-var hasServerSideApply = true
+var (
+	log = logf.Log
+
+	hasServerSideApply atomic.Value
```
Comment (zregvart) [inline pkg/install/kamelets.go:58]:
```suggestion
	hasServerSideApply atomic.Value
	hasServerSideApplyWrite sync.Mutex
```

File: pkg/install/kamelets.go [hunk 3]
```diff
@@ -121,13 +131,14 @@ func applyKamelet(ctx context.Context, c client.Client, path string, namespace s
 	kamelet.GetLabels()[v1alpha1.KameletBundledLabel] = "true"
 	kamelet.GetLabels()[v1alpha1.KameletReadOnlyLabel] = "true"
 
-	if hasServerSideApply {
+	if v := hasServerSideApply.Load(); v.(bool) {
```
Comment (zregvart) [inline pkg/install/kamelets.go:134]:
```suggestion
	if v := hasServerSideApply.Load(); v == nil || v.(bool) {
		if v == nil {
			hasServerSideApplyWrite.Lock()
			defer hasServerSideApplyWrite.Unlock()
		}
```

File: pkg/install/kamelets.go [hunk 4]
```diff
@@ -121,13 +131,14 @@ func applyKamelet(ctx context.Context, c client.Client, path string, namespace s
...
 		err := serverSideApply(ctx, c, kamelet)
 		switch {
 		case err == nil:
 			return nil
```
Comment (zregvart) [inline pkg/install/kamelets.go:138]:
```suggestion
			if v == nil {
				hasServerSideApply.Store(true)
			}
			return nil
```

**Review by zregvart — COMMENTED** [review]
Something like this using a mutex to lock on the first write to `hasServerSideApply` in order to prevent multiple gorutines competing to write to it. Written from the top of my head, might not compile...

Feel free to disregard the whole idea if you think it complicates the implementation needlessly. This will have (perhaps a slight) impact on startup performance, as all goroutines will wait for the first one to complete before running.

**Comment by astefanutti** [comment]
@zregvart I've taken on your idea to use a mutex, and I've gone for using `sync.Once`, which encapsulates the mutex, and exposes just the right semantic for doing the _SSA attempt / fallback_ logic only once.

Interestingly, the installation of the 137 bundled Kamelets now takes less than 200ms on my computer, which is less than what I observed without the extra synchronisation. That being said, we're probably inside the measurement margin error to draw any conclusion.

**Review by zregvart — APPROVED** [review]
Sweet!

**Pull request merged**